src/Orchard.Web/Modules/Orchard.OpenId/Module.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using Orchard.Environment.Shell;
+using Orchard.Security;
+using Orchard.OpenId.Indexes;
+using Orchard.OpenId.Models;
+using Orchard.Users.Models;
+using Orchard.OpenId.Services;
+using Orchard.FileSystem.AppData;
+using Microsoft.Extensions.Logging;
+using Orchard.Settings;
+
+namespace Orchard.OpenId
+{
+    public class Startup : StartupBase
+    {     
+        private readonly string _certificateFullPath;
+        private const string certificateFileName = "Certificate.pfx";
+        private readonly string tenant;
+
+        public Startup(ShellSettings shellSettings, IAppDataFolder appDataFolder, ILoggerFactory loggerFactory)   
+        {
+            tenant = shellSettings.Name;
+            _certificateFullPath = appDataFolder.Combine(shellSettings.Name, certificateFileName);
+        }
+
+        public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
+        {
+            var baseUrl = serviceProvider.GetRequiredService<ISiteService>().GetSiteSettingsAsync().Result.BaseUrl ?? tenant;
+            builder.UseOpenIddict();
```

Comment:
FWIW, OpenIddict has not been designed with "true multi-tenancy in mind" (because well, the MT story in ASP.NET Core is completely inexistent :smile:).

That said, you should be able to provide complete isolation by using different OpenIddict middleware instances, with separate endpoint paths. You'll also have to override the default data protector via `Configure(options => options.DataProtectionProvider = ...)` to ensure tokens and encrypted and HMACed using tenant-specific "purposes strings". Same thing for the signing credentials (for identity tokens and access tokens if you use JWT), that will have to be configured per-tenant.

> You can see an example in the USers module when the authentication cookie is configured, and its path is limited to the tenant name in the url.

I'm a bit surprised the Users module doesn't guarantee such an isolation. Setting a "path per tenant" is not enough to prevent a bad guy from modifying a cookie to make it accepted on a different tenant.